File: non_constructible_change/change.py
# Complexity O(nlog(n)) nlog(n) - complexity of the best sorting algorithm (Heap/Merge Sort),
# Space O(1)
def nonConstructibleChange(coins):
    if len(coins) > 0:
        coins.sort()
        max_change = 0
        for coin in coins:
            if max_change + 1 < coin:
                break
            max_change += coin
        return max_change + 1
    return 1


if __name__ == '__main__':
    print(nonConstructibleChange([1, 1, 1, 1, 1, 1]))

File: non_constructible_change/change.go
package main

import (
	"fmt"
	"sort"
)

func calculateChange(coins *[]int) int {
	var max_change int = 0

	if len(*coins) > 0 {
		sort.Ints(*coins)
		for _, coin := range *coins {
			if max_change+1 < coin {
				break
			}
			max_change += coin
		}
	}
	return max_change + 1
}

//var coins []int = []int{1, 1, 1, 1, 1, 1}
var coins []int = []int{5, 7, 1, 1, 2, 3, 22}

func main() {
	fmt.Println(calculateChange(&coins))
}
